repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/urinormalization/UriNormalizationFilterTest.groovy

```diff
@@ -296,4 +332,20 @@ class UriNormalizationFilterTest extends ReposeValveTest {
         mc.receivedResponse.headers.findAll("via").size() == 1
     }
 
+    // This is a hack to avoid applying configs more than once, which saves on waiting time.
+    //
+    // Some test methods in this class with multiple iterations test against the same repose config file.
+    // On the first iteration of a test method, the config file is reloaded and tests wait for repose to
+    // reload. Since the config file is static, there's no point reapplying and waiting again. So we've
+    // got this helper to only apply configs and wait on the first iteration.
+    static def appliedConfigs = []
+    def applyConfigsOnlyFirstTime(String path, Map params, Closure closure) {
```

Comment:
This method makes me nervous. It relies on test execution order. If, for example, the first test applies a set of configs, the second test applies a different set of configs, and then a third test requires the set of configs applied by the first test, the configs are not reapplied and we have an issue. So my questions are:

- Does Spock ensure that test methods are run in any particular order? What if the `@Unroll` annotation is used?
- Does Spock support parallel test execution within the same test class (or, really, within a `Specification`)?

We probably don't care too much about parallel test execution within a `Specification`. We already have reasons prevent us from doing that, and I question how valuable it would be anyway. Run order is a concern, though. This `Specification` is running fine now, but I worry that upgrading Spock could change that.

You could check the contents of the configuration files rather than their past application using a checksum function, then conditionally apply new configs if they don't match the currently applied configs.

Or to reduce complexity, you could just separate this class into multiple `Specification`s (i.e., `Class`es). You could even keep all of these tests in the same test file if you wanted, though you don't have to. Each `Specification` would apply the configs it needs for the tests it runs. This is basically the approach we use for all of our tests, so it's a natural choice, the only new information I am providing is that I believe it is possible to keep multiple `Specification`s in the same file.

Reply:
Yeah, agreed. 

There's only one method in this class that actually uses multiple iterations (i.e. multiple entries in the `where` table), so I think I'll move that into another class, and leave the others as-is (after removing the `applyConfigsOnlyFirstTime`).